zapcore/write_syncer.go

```diff
@@ -75,6 +77,75 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	bufferWriter *bufio.Writer
+	ws           WriteSyncer
+}
+
+// defaultBufferSize sizes the buffer associated with each WriterSync.
+const defaultBufferSize = 256 * 1024
+
+// defaultFlushInterval means the default flush interval
+const defaultFlushInterval = 30 * time.Second
+
+// Buffer wraps a WriteSyncer in a buffer to improve performance,
+// if bufferSize = 0, we set it to defaultBufferSize
+// if flushInterval = 0, we set it to defaultFlushInterval
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) WriteSyncer {
+	if lws, ok := ws.(*lockedWriteSyncer); ok {
+		if _, ok := lws.ws.(*bufferWriterSyncer); ok {
+			// no need to layer on another buffer
+			return ws
+		}
+	}
+
+	if bufferSize == 0 {
+		bufferSize = defaultBufferSize
+	}
+
+	if flushInterval == 0 {
+		flushInterval = defaultFlushInterval
+	}
+
+	// bufio is not goroutine safe, so add lock writer here
+	ws = Lock(&bufferWriterSyncer{
+		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
+	})
+
+	// flush buffer every interval
+	// we do not need exit this goroutine explicitly
```

Comment:
I think we should add a `Close` method to this interface and allow the `Close` to stop goroutines. Otherwise, we'll leak a goroutine which will affect tests that use something like goleak: https://github.com/uber-go/goleak

We enable goleak for many of our internal tests to verify there's no leaks, and usage of `Buffer` would cause a leak that can't be stopped.

The `Close` method should:
 - do one last flush
 - stop the flush goroutine

Reply:
I agree, and this change may lead to add `Close` method for zap.
Please check the latest commit, thanks.